Update Add Token guides and add bridging guide (#92)
LiskHQ/lisk-docs: docs/building-on-lisk/add-token-to-lisk/custom-token.mdx

---
title: Deploying a custom token
slug: /building-on-lisk/add-token-to-lisk/custom-token
description: 'Learn how to bridge your custom ERC-20 token to Lisk using the standard bridge.'
keywords:
  [
    'ERC-20 contract',
    'Custom token',
    'Lisk Testnet',
    'Sepolia',
    'Ethereum',
    'Lisk Sepolia',
    'Optimism Superchain token list',
  ]
---

> # Deploying Your Custom ERC-20 Token to Lisk

Reviewer: ```suggestion
# Deploying your Custom ERC-20 token to Lisk
```

:::info

> **This tutorial is for developers who want to bridge a new ERC-20 token to Lisk Mainnet.**

Reviewer: ```suggestion
**This tutorial is for developers who want to bridge a new Custom ERC-20 token to Lisk Mainnet.**
```

{/* If you want to bridge existing tokens, you can skip to the tutorial on [Bridging ERC-20 tokens to Lisk](../bridge-tokens-to-lisk). */}

>  If you want to bridge existing tokens, you can skip to the tutorial on [Bridging ERC-20 tokens with the Optimism SDK](https://docs.optimism.io/builders/app-developers/tutorials/cross-dom-bridge-erc20).

Reviewer: ```suggestion
 If you want to bridge existing tokens, you can follow the tutorial on [Bridging ERC-20 tokens with the Optimism SDK](https://docs.optimism.io/builders/app-developers/tutorials/cross-dom-bridge-erc20).
```

:::

In this tutorial you'll learn how to bridge a custom ERC-20 token from Ethereum to Lisk Sepolia using the [Standard Bridge system](https://docs.optimism.io/builders/dapp-developers/bridging/standard-bridge).
This tutorial is meant for developers who already have an existing ERC-20 token on Ethereum and want to create a bridged representation of that token on Lisk Sepolia.

Learn step-by-step how you can create a custom token that conforms to the [`IOptimismMintableERC20`](https://github.com/ethereum-optimism/optimism/blob/v1.1.4/packages/contracts-bedrock/src/universal/IOptimismMintableERC20.sol) interface so that it can be used with the Standard Bridge system.
A custom token allows you to do things like trigger extra logic whenever a token is deposited.

> If you don't need extra functionality like this, consider following the tutorial on [Adding Your Standard ERC-20 Token to Lisk](./standard-token) instead.

Reviewer: ```suggestion
If you don't need extra functionality like this, consider following the tutorial on [Deploying your Standard ERC-20 token to Lisk](./standard-token) instead.
```

## Prerequisites

### Get ETH on Sepolia and Lisk Sepolia

You will need to get some ETH on both, Sepolia and Lisk Sepolia networks.

:::info

> You can use [this faucet](https://sepoliafaucet.com/) to get ETH on Sepolia.

Reviewer: ```suggestion
You can use [ETH Sepolia Faucet](https://sepoliafaucet.com/) to get ETH on Sepolia.
```

You can use the [Superchain Faucet](https://app.optimism.io/faucet?utm_source=docs) to get ETH on Lisk Sepolia.
:::

### Add Lisk Sepolia to Your Wallet

This tutorial uses [Remix](https://remix.ethereum.org) to deploy contracts.
You will need to add the Lisk Sepolia network to your wallet in order to follow this tutorial.

> Please follow [this guide](../connecting-to-a-wallet.mdx) to connect your wallet to Lisk Sepolia.

Reviewer: ```suggestion
Please follow the [How to connect Lisk to a wallet](../connecting-to-a-wallet.mdx) guide, to connect your wallet to Lisk Sepolia.
```

        override(IOptimismMintableERC20)
        onlyBridge
    {
        _mint(_to, _amount);
        emit Mint(_to, _amount);
    }

    /// @notice Prevents tokens from being withdrawn to L1.
    function burn(
        address,
        uint256
    )
        external
        virtual
        override(IOptimismMintableERC20)
        onlyBridge
    {
        revert("MyCustomL2Token cannot be withdrawn");
    }
}
```
</details>

### 4. Review the example contract

Take a moment to review the example contract.
It's almost the same as the standard [`OptimismMintableERC20`](https://github.com/ethereum-optimism/optimism/blob/v1.1.4/packages/contracts-bedrock/src/universal/OptimismMintableERC20.sol) contract except that the `_burn` function has been made to always revert.

The contract for the custom token inherits from the [`IOptimismMintableERC20`](https://github.com/ethereum-optimism/optimism/blob/v1.1.4/packages/contracts-bedrock/src/universal/IOptimismMintableERC20.sol) interface and the `ERC20` contract.
The constructor takes the address of the L2 standard bridge, the address of the corresponding L1 token, the name of the ERC20 token, and the symbol of the ERC20 token.

> The `mint` function allows the bridge to mint tokens to users.

Reviewer: ```suggestion
The `mint` function allows the bridge to mint tokens for users.
```

Since the bridge needs to burn tokens when users want to withdraw them to L1, this means that users will not be able to withdraw tokens from this contract, which is what we intend for this example.

```solidity
/// @notice Prevents tokens from being withdrawn to L1.
function burn(
    address,
    uint256
)
    external
    virtual
    override(IOptimismMintableERC20)
    onlyBridge
{
    revert("MyCustomL2Token cannot be withdrawn");
}
```

### 5. Compile the contract

Save the file to automatically compile the contract.

> If you've disabled auto-compile, you'll need to manually compile the contract by clicking the "Solidity Compiler" tab (this looks like the letter "S") and press the blue "Compile" button.

Reviewer: ```suggestion
If you've disabled auto-compile, you'll need to manually compile the contract by clicking the "Solidity Compiler" tab (this looks like the letter "S") and pressing the blue "Compile" button.
```
